packages/epetra/test/IntMultiVectorDistributed/cxx_main.cpp

```diff
@@ -0,0 +1,183 @@
+#include "Epetra_BlockMap.h"
+#include "Epetra_CombineMode.h"
+#include "Epetra_Import.h"
+#ifdef EPETRA_MPI
+#include "Epetra_MpiComm.h"
+#include <mpi.h>
+#else
+#include "Epetra_SerialComm.h"
+#endif
+#include "../epetra_test_err.h"
+#include "Epetra_Version.h"
+
+int main(int argc, char *argv[]) {
+  int ierr = 0;
+  int global_ierr = 0;
+  int numErr = 0;
+
+#ifdef EPETRA_MPI
+
+  // Initialize MPI
+
+  MPI_Init(&argc,&argv);
+  int rank; // My process ID
+
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  Epetra_MpiComm Comm( MPI_COMM_WORLD );
+
+#else
+
+  int rank = 0;
+  Epetra_SerialComm Comm;
+
+#endif
+
+  bool verbose = false;
+
+  // Check if we should print results to standard out
+  if(argc > 1) {
+    if(argv[1][0]=='-' && argv[1][1]=='v') {
+      verbose = true;
+    }
+  }
+
+  Comm.SetTracebackMode(0); // This should shut down any error traceback reporting
+  int MyPID = Comm.MyPID();
+  int NumProc = Comm.NumProc();
+
+  if(verbose && MyPID==0)
+    cout << Epetra_Version() << endl << endl;
+
+  if(verbose) cout << "Processor " << MyPID << " of " << NumProc << " is alive." << endl;
+
+  // Redefine verbose to only print on PE 0
+  if(verbose && rank != 0)
+		verbose = false;
+
+  int NumMyEquations = 6;
+  int NumGlobalEquations = NumMyEquations*NumProc+EPETRA_MIN(NumProc,3);
+  if(MyPID < 3)
+    NumMyEquations++;
+
+  // Construct a Map that puts approximately the same Number of equations on each processor
+
+  Epetra_BlockMap& Map = *new Epetra_BlockMap(NumGlobalEquations, NumMyEquations, 1, 0, Comm);
+
+  // Get update list and number of local equations from newly created Map
+  int* MyGlobalElements = new int[Map.NumMyElements()];
+  Map.MyGlobalElements(MyGlobalElements);
+
+  // Construct the IntMultiVector
+  const int numVecs = 3;
+  Epetra_IntMultiVector& myIMV = *new Epetra_IntMultiVector(Map, numVecs);
+
+  // Extract a view to populate the IntMultiVector
+  int myLDA;
+  int* myVals;
+  ierr = myIMV.ExtractView(&myVals, &myLDA);
+  Comm.MaxAll(&ierr, &global_ierr, 1);
+  EPETRA_TEST_ERR(global_ierr, numErr);
+  for(int i = 0; i < myIMV.MyLength(); ++i) {
+    for(int j = 0; j < numVecs; ++j) {
+      myVals[j*myLDA + i] = numVecs*i + j;
+    }
+  }
+
+  // Test: replace local values
+  myIMV.ReplaceMyValue(3, 2, 42);
+  ierr = (myIMV[2][3] == 42 ? 0 : 1);
+  Comm.MaxAll(&ierr, &global_ierr, 1);
+  EPETRA_TEST_ERR(global_ierr, numErr);
+
+  // Test: sum into local values
+  myIMV.SumIntoMyValue(3, 1, 42);
+  ierr = (myIMV[1][3] == 3*numVecs + 1 + 42 ? 0 : 1);
+  Comm.MaxAll(&ierr, &global_ierr, 1);
+  EPETRA_TEST_ERR(global_ierr, numErr);
+
+  // Test: replace and sum global values on rank 2
+  if(MyPID == 2) {
+    myIMV.ReplaceGlobalValue(Map.GID(3), 2, 48);
+    ierr = (myIMV[2][3] == 48 ? 0 : 1);
+    EPETRA_TEST_ERR(ierr, numErr);
+    myIMV.SumIntoGlobalValue(Map.GID(3), 1, 48);
+    ierr = (myIMV[1][3] == 3*numVecs + 1 + 42 + 48 ? 0 : 1);
+    EPETRA_TEST_ERR(ierr, numErr);
+  }
```

Comment:
It would be nicer to put one of those `Comm.MaxAll` things right after the `MyPID == 2` block, rather than relying on `Comm.MaxAll` calls below.  However, you do at least have a `Comm.MaxAll` call below.

Reply:
@mhoemmen this piece does not use `Comm.MaxAll` because it only works on rank 2, should we test that on the other ranks the value in `myIMV[1][3]` has not changed?
Or do you want to call `Comm.MaxAll` even though `ierr` cannot be modified on ranks 0, 1 and 3?

Reply:
"right after [i.e., outside] the `MyPID == 2` block" -- that way it's easier to catch errors sooner.  This only matters, though, if you print errors and/or exit right away after `Comm.MaxAll`, so it's not a big deal.